ding_unilogin/src/Controller/InstitutionController.php

```diff
@@ -0,0 +1,114 @@
+<?php
+
+namespace Drupal\ding_unilogin\Controller;
+
+use Drupal\ding_unilogin\Exception\HttpBadRequestException;
+use Drupal\ding_unilogin\Exception\HttpNotFoundException;
+use Drupal\ding_unilogin\Exception\HttpUnauthorizedException;
+
+/**
+ * Institution controller.
+ */
+class InstitutionController {
+
+  /**
+   * Handle request.
+   */
+  public function handle(array $args) {
+    $authorization = $this->getRequestHeader('authorization');
+    if (!preg_match('/^(bearer|token) (?P<token>.+)$/i', $authorization, $matches)) {
+      throw new HttpUnauthorizedException();
+    }
+    $token = $matches['token'];
+
+    if (empty($args)) {
+      $method = $_SERVER['REQUEST_METHOD'];
+
+      if ('POST' === $method) {
+        if (variable_get('ding_unilogin_api_token_write') !== $token) {
+          throw new HttpUnauthorizedException();
+        }
+
+        return $this->update();
+      }
+
+      if (variable_get('ding_unilogin_api_token_read') !== $token) {
+        throw new HttpUnauthorizedException();
+      }
+
+      return $this->list();
+    }
+    elseif (1 === \count($args)) {
+      return $this->read($args[0]);
+    }
+
+    throw new HttpBadRequestException('Invalid request');
+  }
+
+  /**
+   * List of institutions.
+   */
+  public function list() {
+    $institutions = _ding_unilogin_get_institutions();
+
+    return ['data' => $institutions];
+  }
+
+  /**
+   * Get an institution by id.
+   */
+  public function read($id) {
+    $institutions = _ding_unilogin_get_institutions();
+
+    if (isset($institutions[$id])) {
+      return ['data' => $institutions[$id]];
+    }
+
+    throw new HttpNotFoundException(sprintf('Invalid institution id: %s', $id));
+  }
+
+  /**
+   * Update the list of institutions.
+   */
+  public function update() {
```

Comment:
Is this not more a `create` as it comes through a `POST` request and overrides all the institutions? Should an `PATCH` have been create to update a single institutions in the list?

Reply:
It is an update (actually an overwrite via `POST`). We don't need to be able to update a single institution (for now).